ff/serving/provider/postgres.go

```diff
@@ -0,0 +1,501 @@
+		return nil, errors.New("only features can be materialized")
+	}
+	resTable, err := store.getPostgresResourceTable(id) // Need to clarify names
+	if err != nil {
+		return nil, err
+	}
+
+	matID := MaterializationID(id.Name)
+	matTableName := store.getMaterializationTableName(matID)
+	sanitizedTableName := sanitize(matTableName)
+	resTableName := sanitize(resTable.name)
+	tableCreateQry := fmt.Sprintf("CREATE TABLE IF NOT EXISTS %s AS (SELECT entity, value, ts FROM %s WHERE 1=2)", sanitizedTableName, resTableName)
+
+	_, err = store.conn.Exec(
+		context.Background(), tableCreateQry) // Set correct types
+	if err != nil {
+		return nil, err
+	}
+
+	materializeQry := fmt.Sprintf(
+		"INSERT INTO %s SELECT entity, value, ts FROM "+
+			"(SELECT entity, ts, value, row_number() OVER (PARTITION BY entity ORDER BY ts desc) "+
+			"AS rn FROM %s) t WHERE rn=1", sanitizedTableName, resTableName)
+
+	_, err = store.conn.Exec(context.Background(), materializeQry)
+	if err != nil {
+		return nil, err
+	}
+
+	return &postgresMaterialization{
+		id:        matID,
+		conn:      store.conn,
+		tableName: matTableName,
+	}, nil
+
+}
+func (store *postgresOfflineStore) GetMaterialization(id MaterializationID) (Materialization, error) {
+	getMatQry := fmt.Sprintf("SELECT DISTINCT (table_name) FROM information_schema.tables WHERE table_name LIKE 'materialization_%s%%'", string(id))
+	rows, err := store.conn.Query(context.Background(), getMatQry)
+	defer rows.Close()
+	if err != nil {
+		return nil, err
+	}
+	rowCount := 0
+	var tableName string
+	for rows.Next() {
+		if err := rows.Scan(&tableName); err != nil {
+			return nil, err
+		}
+		rowCount++
+	}
+	if rowCount == 0 {
+		return nil, &MaterializationNotFound{id}
+	}
+	return &postgresMaterialization{
+		id:        id,
+		conn:      store.conn,
+		ctx:       store.ctx,
+		tableName: tableName,
+	}, err
+}
+
+func (store *postgresOfflineStore) CreateTrainingSet(def TrainingSetDef) error {
+	if err := def.check(); err != nil {
+		return err
+	}
+	label, err := store.getPostgresResourceTable(def.Label)
+	if err != nil {
+		return err
+	}
+	tableName := store.getTrainingSetName(def.ID)
+
+	columns := make([]string, 0)
+	query := fmt.Sprintf(" (SELECT entity, value, ts from %s ) l ", sanitize(label.name))
+	for i, feature := range def.Features {
+		resourceTableName := sanitize(store.getResourceTableName(feature))
+		tableJoinAlias := fmt.Sprintf("t%d", i)
+		columns = append(columns, resourceTableName)
+		query = fmt.Sprintf("%s LEFT JOIN LATERAL (SELECT entity , value as %s, ts  FROM %s WHERE entity=l.entity and ts <= l.ts ORDER BY ts desc LIMIT 1) %s on %s.entity=l.entity ",
+			query, resourceTableName, resourceTableName, tableJoinAlias, tableJoinAlias)
+		if i == len(def.Features)-1 {
+			query = fmt.Sprintf("%s )", query)
+		}
+	}
+	columnStr := strings.Join(columns, ", ")
+	fullQuery := fmt.Sprintf("CREATE TABLE %s AS (SELECT %s, l.value  FROM %s ", sanitize(tableName), columnStr, query)
```

Comment:
l.value should be aliased as the label resource table name

Reply:
Done